range.c

```diff
@@ -1017,6 +1017,30 @@ range_to_a(VALUE range)
     return rb_call_super(0, 0);
 }
 
+/*
+ *  call-seq:
+ *    to_a -> set
+ *
+ *  Returns a set containing the elements in +self+, if a finite collection;
+ *  raises an exception otherwise.
+ *
+ *    (1..4).to_set     # => Set[1, 2, 3, 4]
+ *    (1...4).to_set    # => Set[1, 2, 3]
+ *    ('a'..'d').to_set # => Set["a", "b", "c", "d"]
+ *    ('a'..).to_set # => raises RangeError
+ *
+ */
+
+
+static VALUE
+range_to_set(VALUE range)
+{
+    if (NIL_P(RANGE_END(range))) {
+        rb_raise(rb_eRangeError, "cannot convert endless range to a set");
```

Comment:
I believe that endless range error should be the same error type as beginless range error.  Calling `to_set` on a beginless range currently raises `TypeError`.  This method should probably have beginless range raise `RangeError` instead, and include a test and an example for that.

Reply:
Makes sense. Should we also tweak `Range#to_a` to raise `RangeError` in both cases?

```shell
./ruby -e '(..1).to_a'
-e:1:in 'Range#each': can't iterate from NilClass (TypeError)
```

Reply:
It feels clearer for me.
@knu Any thoughts?

If we will change it too, maybe ruby/spec also needs to be updated?

Reply:
We should perhaps check the size of a given Enumerable in Set.new and raise an error when it is infinity.

Reply:
> We should perhaps check the size of a given Enumerable in Set.new and raise an error when it is infinity.

Seems good to me. I'll try to follow this path

Reply:
I tried out this idea and it seems to work fine (we now raise `ArgumentError` on error infinite enumerable, e.g. `1.upto(Float::INFINITY)`.

```shell
$ ./ruby -e 'Set.new((1..nil))'
-e:1:in 'Set#initialize': cannot initialize Set from a enumerable with infinite size (ArgumentError)
	from -e:1:in '<main>'
$ ./ruby -e '(1..nil).to_set'  
-e:1:in 'Enumerable#to_set': cannot initialize Set from a enumerable with infinite size (ArgumentError)
	from -e:1:in '<main>'

```